### PR TITLE
cli: use clap for cmdline arg parsing

### LINKED ISSUE #15: Improve CLI tool argument parsing
Currently the CLI tool `npc` only takes interface name for filtering with json output only.

To improve it:

- [ ] `--yaml` argument for showing in yaml format
- [ ] Support `npc route dev <interfae_name>` for showing routes next hop to specified interface.
- [ ] Support `npc route table <table_id>` for showing routes of specified route table.
- [ ] Support `npc route type l` for showing link local routes. And `u` for unicast, etc. Please refer to `RouteType`.

### PR DESCRIPTION
**Pull request by acardace**
Now you have subcommands 'link', 'route', 'help' and '-V' and '-h' switches.

By using clap you also easily get a usage message and everything else
you would expect from a cli app.

Signed-off-by: Antonio Cardace <acardace@redhat.com>

### CODE REVIEW

File: src/lib/net_state.rs [hunk 1]
```diff
@@ -12,8 +12,10 @@ pub struct NetState {
...
-pub fn get_state() -> Result<NetState, NisporError> {
-    let ifaces = get_ifaces()?;
-    let routes = get_routes(&ifaces)?;
-    Ok(NetState { ifaces, routes })
+impl NetState {
+    pub fn new() -> Result<NetState, NisporError> {
+        let ifaces = get_ifaces()?;
+        let routes = get_routes(&ifaces)?;
+        Ok(NetState { ifaces, routes })
+    }
```
Comment (cathay4t) [inline src/lib/net_state.rs]: This will break API and a lot need to be changed: C and Python binding, test cases.

Please move this patch into separate PR for discussion.

File: src/cli/npc.rs [hunk 2]
```diff
@@ -1,35 +1,59 @@
...
+use std::process;
 
 fn main() {
-    let argv: Vec<String> = args().collect();
-    match get_state() {
+    let matches = clap_app!(npc =>
+        (version: crate_version!())
+        (author: crate_authors!())
+        (about: "Nispor CLI")
+        (@subcommand link =>
```
Comment (cathay4t) [inline src/cli/npc.rs]: Can we use `npc eth1` instead of `npc link eth1`?

And assuming there is no interface named as `route`, so `npc route dev eth1` is acceptable.

This might require some hack in clap, but that could save a lot typing time of me.

File: src/cli/npc.rs [hunk 3]
```diff
@@ -1,35 +1,59 @@
...
+    };
+
+    match result {
+        Ok(json) => {
+            println!("{}", json);
+            process::exit(0);
+        }
         Err(e) => {
             eprintln!("{}", e);
+            process::exit(1);
```
Comment (cathay4t) [inline src/cli/npc.rs:89]: How about we print error in json format also?

**Review by cathay4t — CHANGES_REQUESTED** [review]

File: src/lib/net_state.rs (hunk 1 above)
Comment (acardace) [reply src/lib/net_state.rs]: Moved to https://github.com/nispor/nispor/pull/27

File: src/cli/npc.rs (hunk 2 above)
Comment (acardace) [reply src/cli/npc.rs]: Are you sure? Having 'npc IFNAME' and 'npc route IFNAME' seems a bit inconsistent.

File: src/cli/npc.rs (hunk 3 above)
Comment (acardace) [reply src/cli/npc.rs:89]: How about now?

File: src/cli/npc.rs (hunk 2 above)
Comment (cathay4t) [reply src/cli/npc.rs]: It might be tricky for code and document. let me think it through.

**Comment by cathay4t** [comment]
The most of my time using `npc`, I use it to query interface. So there is no need for me to type `npc iface` or `npc link` again and again.

So what I expect from this PR:
 * `npc`: dump everything.
 * `npc eth1`: interface information of eth1.
 * `npc route`: all route information. This assume there will never a interface call route. But I can live with that as this is just a cli tool designed for debuging.

For the remaining, please refer to https://github.com/nispor/nispor/issues/15

**Comment by cathay4t** [comment]
@ichigo663 Hi Antonio,

I like your on-going approach. Do you need any help  on getting it done?

Few suggestions:
 * Only use single line/function converting struct to serialize(to json), all pre-print functions only do struct level filtering.
   And the serilize function could handle json/yaml/toml without understanding of filtering.
 * Add `Serialize` derive to `NisporError`, pre-print functions could `NisporCliError` with `Serialize` enabled. So the later serilize function could just showing the error into expect format(json/yaml/toml).

File: src/cli/npc.rs [hunk 4]
```diff
@@ -1,41 +1,92 @@
...
+                CliResult::Full(state)
             }
         }
-        Err(e) => {
-            eprintln!("{}", e);
+        Err(e) => CliResult::NisporError(e),
+    };
+
+    match result {
+        CliResult::Full(netstate) => {
```
Comment (acardace) [inline src/cli/npc.rs:67]: If anyone can write a macro for this it'd be great, I'm not really that skilled in rust ;)

Comment (cathay4t) [reply src/cli/npc.rs:67]:
```rust
match result {
    CliResult::Error(e) | CliResult::NisporError(e) => something,
    _ => println!()
}
```

**Review by cathay4t — APPROVED** [review]

File: src/cli/npc.rs (hunk 4 above)
Comment (acardace) [reply src/cli/npc.rs:67]: Unfortunately that does not work, all binding must have the same type in all alternatives.
